File: python/0020-valid-parentheses.py
class Solution(object):
    def isValid(self, s):
        brackets_map = {')': '(', ']': '[', '}': '{'}
        stack = []

        for c in s:
            if c in brackets_map:
                if not stack or stack[-1] != brackets_map[c]:
                    return False

                stack.pop()
            else:
                stack.append(c)

        return not stack

File: javascript/0020-valid-parentheses.js
var isValid = function(s) {
    const map = {')': '(', '}': '{', ']': '['}
    const stack = []
    for(const c of s) {
        if(c in map) {
            if(!stack.length || stack[stack.length - 1] !== map[c]) return false
            stack.pop()
        }
        else stack.push(c)
    }
    return !stack.length
};
